ccl/kvccl/kvtenantccl: TestTenantUpgrade failed

ccl/kvccl/kvtenantccl.TestTenantUpgrade [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/8691945?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/8691945?buildTab=artifacts#/) on master @ [8511dc08269679b6562f2010ef357c2384649871](https://github.com/cockroachdb/cockroach/commits/8511dc08269679b6562f2010ef357c2384649871):

Fatal error:

```
panic: test timed out after 4m55s

```
Stack: 

```
goroutine 22439 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2036 +0x88
created by time.goFunc
	GOROOT/src/time/sleep.go:176 +0x38
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestTenantUpgrade
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/c33cccbcd49aa55ca01529df00fb9d50/logTestTenantUpgrade2115985398
    test_log_scope.go:79: use -show-logs to present logs inline
=== RUN   TestTenantUpgrade/upgrade_tenant
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/multi-tenant
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTenantUpgrade.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-24493

Epic CRDB-20293

ccl/kvccl/kvtenantccl.TestTenantUpgrade [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/8702182?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/8702182?buildTab=artifacts#/) on master @ [f3ff41774a902d6005dbfad504135e64d9434daf](https://github.com/cockroachdb/cockroach/commits/f3ff41774a902d6005dbfad504135e64d9434daf):

Fatal error:

```
panic: test timed out after 4m55s

```
Stack: 

```
goroutine 23237 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2036 +0x88
created by time.goFunc
	GOROOT/src/time/sleep.go:176 +0x38
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestTenantUpgrade
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/c33cccbcd49aa55ca01529df00fb9d50/logTestTenantUpgrade2165436414
    test_log_scope.go:79: use -show-logs to present logs inline
=== RUN   TestTenantUpgrade/upgrade_tenant
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTenantUpgrade.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/kvccl/kvtenantccl.TestTenantUpgrade [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/8702182?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/8702182?buildTab=artifacts#/) on master @ [f3ff41774a902d6005dbfad504135e64d9434daf](https://github.com/cockroachdb/cockroach/commits/f3ff41774a902d6005dbfad504135e64d9434daf):

Fatal error:

```
panic: test timed out after 4m55s

```
Stack: 

```
goroutine 21962 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2036 +0x88
created by time.goFunc
	GOROOT/src/time/sleep.go:176 +0x38
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestTenantUpgrade
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/c33cccbcd49aa55ca01529df00fb9d50/logTestTenantUpgrade3312468972
    test_log_scope.go:79: use -show-logs to present logs inline
=== RUN   TestTenantUpgrade/upgrade_tenant
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTenantUpgrade.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/kvccl/kvtenantccl.TestTenantUpgrade [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/8708528?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/8708528?buildTab=artifacts#/) on master @ [186fa03cea267357d7a7dec31a815e8b853d6949](https://github.com/cockroachdb/cockroach/commits/186fa03cea267357d7a7dec31a815e8b853d6949):

Fatal error:

```
panic: test timed out after 4m55s

```
Stack: 

```
goroutine 22212 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2036 +0x88
created by time.goFunc
	GOROOT/src/time/sleep.go:176 +0x38
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestTenantUpgrade
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/c33cccbcd49aa55ca01529df00fb9d50/logTestTenantUpgrade1156790019
    test_log_scope.go:79: use -show-logs to present logs inline
=== RUN   TestTenantUpgrade/upgrade_tenant
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTenantUpgrade.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/kvccl/kvtenantccl.TestTenantUpgrade [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/8708528?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/8708528?buildTab=artifacts#/) on master @ [186fa03cea267357d7a7dec31a815e8b853d6949](https://github.com/cockroachdb/cockroach/commits/186fa03cea267357d7a7dec31a815e8b853d6949):

Fatal error:

```
panic: test timed out after 4m55s

```
Stack: 

```
goroutine 23085 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2036 +0x88
created by time.goFunc
	GOROOT/src/time/sleep.go:176 +0x38
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestTenantUpgrade
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/c33cccbcd49aa55ca01529df00fb9d50/logTestTenantUpgrade3142133801
    test_log_scope.go:79: use -show-logs to present logs inline
=== RUN   TestTenantUpgrade/upgrade_tenant
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTenantUpgrade.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/kvccl/kvtenantccl.TestTenantUpgrade [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/8709629?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/8709629?buildTab=artifacts#/) on master @ [e73303af7defadebfeb9b02802693e19d5709133](https://github.com/cockroachdb/cockroach/commits/e73303af7defadebfeb9b02802693e19d5709133):

Fatal error:

```
panic: test timed out after 4m55s

```
Stack: 

```
goroutine 22359 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2036 +0x88
created by time.goFunc
	GOROOT/src/time/sleep.go:176 +0x38
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestTenantUpgrade
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/c33cccbcd49aa55ca01529df00fb9d50/logTestTenantUpgrade1857839183
    test_log_scope.go:79: use -show-logs to present logs inline
=== RUN   TestTenantUpgrade/upgrade_tenant
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTenantUpgrade.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/kvccl/kvtenantccl.TestTenantUpgrade [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/8717375?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/8717375?buildTab=artifacts#/) on master @ [5455e9e7e4d533d8ca5dfab496a116888a610c58](https://github.com/cockroachdb/cockroach/commits/5455e9e7e4d533d8ca5dfab496a116888a610c58):

Fatal error:

```
panic: test timed out after 4m55s

```
Stack: 

```
goroutine 22370 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2036 +0x88
created by time.goFunc
	GOROOT/src/time/sleep.go:176 +0x38
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestTenantUpgrade
I230215 14:35:42.579167 1 (gostd) rand.go:199  [T1] 1  random seed: 7966078340390321925
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/c33cccbcd49aa55ca01529df00fb9d50/logTestTenantUpgrade550108376
    test_log_scope.go:79: use -show-logs to present logs inline
=== RUN   TestTenantUpgrade/upgrade_tenant
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTenantUpgrade.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/kvccl/kvtenantccl.TestTenantUpgrade [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/8718438?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/8718438?buildTab=artifacts#/) on master @ [6764f27951db9af6aaa9368aa3fd22aaa8e30e89](https://github.com/cockroachdb/cockroach/commits/6764f27951db9af6aaa9368aa3fd22aaa8e30e89):

Fatal error:

```
panic: test timed out after 4m55s

```
Stack: 

```
goroutine 22066 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2036 +0x88
created by time.goFunc
	GOROOT/src/time/sleep.go:176 +0x38
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestTenantUpgrade
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/c33cccbcd49aa55ca01529df00fb9d50/logTestTenantUpgrade40459336
    test_log_scope.go:79: use -show-logs to present logs inline
=== RUN   TestTenantUpgrade/upgrade_tenant
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTenantUpgrade.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/kvccl/kvtenantccl.TestTenantUpgrade [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/8719385?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/8719385?buildTab=artifacts#/) on master @ [f91cf78f847590e2c2f9902590ea0edc1906537e](https://github.com/cockroachdb/cockroach/commits/f91cf78f847590e2c2f9902590ea0edc1906537e):

Fatal error:

```
panic: test timed out after 4m55s

```
Stack: 

```
goroutine 23088 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2036 +0x88
created by time.goFunc
	GOROOT/src/time/sleep.go:176 +0x38
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestTenantUpgrade
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/c33cccbcd49aa55ca01529df00fb9d50/logTestTenantUpgrade3207185038
    test_log_scope.go:79: use -show-logs to present logs inline
=== RUN   TestTenantUpgrade/upgrade_tenant
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTenantUpgrade.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/kvccl/kvtenantccl.TestTenantUpgrade [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/8724178?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/8724178?buildTab=artifacts#/) on master @ [2cd982f9fd11c6a8c85faccc3cace2d3329ff863](https://github.com/cockroachdb/cockroach/commits/2cd982f9fd11c6a8c85faccc3cace2d3329ff863):

Fatal error:

```
panic: test timed out after 4m55s

```
Stack: 

```
goroutine 22838 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2036 +0x88
created by time.goFunc
	GOROOT/src/time/sleep.go:176 +0x38
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestTenantUpgrade
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/c33cccbcd49aa55ca01529df00fb9d50/logTestTenantUpgrade1860869024
    test_log_scope.go:79: use -show-logs to present logs inline
=== RUN   TestTenantUpgrade/upgrade_tenant
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTenantUpgrade.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/kvccl/kvtenantccl.TestTenantUpgrade [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/8730653?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/8730653?buildTab=artifacts#/) on master @ [af3bf237ce3774258a13a0c2aa0e2da22716270f](https://github.com/cockroachdb/cockroach/commits/af3bf237ce3774258a13a0c2aa0e2da22716270f):

Fatal error:

```
panic: test timed out after 4m55s

```
Stack: 

```
goroutine 22919 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2036 +0x88
created by time.goFunc
	GOROOT/src/time/sleep.go:176 +0x38
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestTenantUpgrade
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/c33cccbcd49aa55ca01529df00fb9d50/logTestTenantUpgrade979347991
    test_log_scope.go:79: use -show-logs to present logs inline
=== RUN   TestTenantUpgrade/upgrade_tenant
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTenantUpgrade.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/kv/kvserver/batcheval/cmd_add_sstable.go
+++ b/pkg/kv/kvserver/batcheval/cmd_add_sstable.go
@@ -77,8 +77,7 @@ var AddSSTableRewriteConcurrency = settings.RegisterIntSetting(
 	settings.SystemOnly,
 	"kv.bulk_io_write.sst_rewrite_concurrency.per_call",
 	"concurrency to use when rewriting sstable timestamps by block, or 0 to use a loop",
-	// TODO(radu): temporarily disabled because of #97076.
-	0, // int64(util.ConstantWithMetamorphicTestRange("addsst-rewrite-concurrency", 4, 0, 16)),
+	int64(util.ConstantWithMetamorphicTestRange("addsst-rewrite-concurrency", 4, 0, 16)),
 	settings.NonNegativeInt,
 )
 

--- a/pkg/storage/sst.go
+++ b/pkg/storage/sst.go
@@ -24,6 +24,7 @@ import (
 	"github.com/cockroachdb/errors"
 	"github.com/cockroachdb/pebble/sstable"
 	"github.com/cockroachdb/pebble/vfs"
+	"github.com/cockroachdb/redact"
 )
 
 var (
@@ -1213,15 +1214,22 @@ func UpdateSSTTimestamps(
 		if fp := defaults.Levels[0].FilterPolicy; fp != nil && len(opts.Filters) == 0 {
 			opts.Filters = map[string]sstable.FilterPolicy{fp.Name(): fp}
 		}
-		if _, err := sstable.RewriteKeySuffixes(sst,
+		rewriteOpts, minTableFormat := makeSSTRewriteOptions(ctx, st)
+		_, tableFormat, err := sstable.RewriteKeySuffixesAndReturnFormat(sst,
 			opts,
 			sstOut,
-			MakeIngestionWriterOptions(ctx, st),
+			rewriteOpts,
 			EncodeMVCCTimestampSuffix(from),
 			EncodeMVCCTimestampSuffix(to),
 			concurrency,
-		); err != nil {
+		)
+		if err != nil {
 			return nil, enginepb.MVCCStats{}, err
+		}
+		if minTableFormat > tableFormat {
+			return nil, enginepb.MVCCStats{},
+				errors.Errorf("rewrite table format %s is less than min format %s",
+					redact.SafeString(tableFormat.String()), redact.SafeString(minTableFormat.String()))
 		}
 		return sstOut.Bytes(), statsDelta, nil
 	}

--- a/pkg/storage/sst_writer.go
+++ b/pkg/storage/sst_writer.go
@@ -75,6 +75,17 @@ func MakeIngestionWriterOptions(ctx context.Context, cs *cluster.Settings) sstab
 	return opts
 }
 
+// makeSSTRewriteOptions should be used instead of MakeIngestionWriterOptions
+// when we are going to rewrite ssts. It additionally returns the minimum
+// table format that we accept, since sst rewriting will often preserve the
+// input table format.
+func makeSSTRewriteOptions(
+	ctx context.Context, cs *cluster.Settings,
+) (opts sstable.WriterOptions, minTableFormat sstable.TableFormat) {
+	// v22.2 clusters use sstable.TableFormatPebblev2.
+	return MakeIngestionWriterOptions(ctx, cs), sstable.TableFormatPebblev2
+}
+
 // MakeBackupSSTWriter creates a new SSTWriter tailored for backup SSTs which
 // are typically only ever iterated in their entirety.
 func MakeBackupSSTWriter(ctx context.Context, cs *cluster.Settings, f io.Writer) SSTWriter {

